source/opt/def_use_manager.h

```diff
@@ -54,17 +54,35 @@ inline bool operator<(const Use& lhs, const Use& rhs) {
   return lhs.operand_index < rhs.operand_index;
 }
 
+// Definition and user pair.
+//
+// The first element of the pair is the definition.
+// The second element of the pair is the user.
+//
+// Definition should never be null. User can be null, however, such an entry
+// should be used only for searching (e.g. all users of a particular definition)
+// and never stored in a container.
 using UserEntry = std::pair<ir::Instruction*, ir::Instruction*>;
 
-struct less_than_id_ptr {
+// Orders UserEntry for use in associative containers (i.e. less than ordering).
+//
+// The definition of an UserEntry is treated as the major key and the users as
+// the minor key so that all the users of a particular definition are
+// consecutive in a container.
+//
+// A null user always compares less than a real user. This is done provide easy
+// values to search for the beginning of the users of a particular definition
+// (i.e. using {def, nullptr}).
+struct UserEntryLess {
   bool operator()(const UserEntry& lhs, const UserEntry& rhs) const {
+    // If lhs.first and rhs.first are both null, fall through to checking the
+    // second entries.
```

Comment:
If the def must never be null, then some of this code is purely defensive.... which is fine.